In [7]:
using JuMP
using GLPK

# Exercise 1 
A house wife wishes to mix two types of food F1 and F2 in such a way that the vitamin contents of the mixture contain at least 8 units of vitamin A and 11 units of vitamin B. Food F1 costs 60€/Kg and Food F2 costs 80€/kg. Food F1 contains 3 units/kg of vitamin A and 5 units/kg of vitamin B while Food F2 contains 4 units/kg of vitamin A and 2 units/kg of vitamin B. The goal is to minimize the cost of the mixtures.

Formulating as a linear programming problem: let $x_1$, $x_2$ be the weights of F1 and F2 in the mix. \
Then: \
min z(x) =    60$x_1$ + 80$x_2$        (0) - minimizing cost of the mixture \
                     3$x_1$ + 4$x_2$ >= 8    (1) - requirement for vitamin A in the mixture \
                     5$x_1$ + 2$x_2$ >= 11   (2) - requirement for vitamin B in the mixture \
                     $x_1$, $x_2$ >= 0           (3) - weights can't be negative 


In [60]:
model1 = Model(GLPK.Optimizer)

A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: GLPK

In [61]:
@variable(model1, x1 >= 0)
@variable(model1, x2 >= 0)

x2

In [62]:
@objective(model1, Min, 60x1 + 80x2)

60 x1 + 80 x2

In [63]:
@constraint(model1, cst1, 3x1 + 4x2 >= 8)
@constraint(model1, cst2, 5x1 + 2x2 >= 11)

cst2 : 5 x1 + 2 x2 ≥ 11

In [64]:
print(model1)

In [65]:
optimize!(model1)

In [66]:
@show termination_status(model1)
@show primal_status(model1)

termination_status(model1) = MathOptInterface.OPTIMAL
primal_status(model1) = MathOptInterface.FEASIBLE_POINT


FEASIBLE_POINT::ResultStatusCode = 1

In [67]:
@show objective_value(model1)
@show value(x1)
@show value(x2)

objective_value(model1) = 160.00000000000003
value(x1) = 2.666666666666667
value(x2) = 0.0


0.0

In [68]:
solution_summary(model1)

* Solver : GLPK

* Status
  Result count       : 1
  Termination status : OPTIMAL
  Message from the solver:
  "Solution is optimal"

* Candidate solution (result #1)
  Primal status      : FEASIBLE_POINT
  Dual status        : FEASIBLE_POINT
  Objective value    : 1.60000e+02
  Objective bound    : -Inf
  Dual objective value : 1.60000e+02

* Work counters
  Solve time (sec)   : 5.00679e-05


## Answer
House wife should use $2\frac{2}{3}$kg of food F1 and no food F2.
The cost of the mixture will be 160 Euros.

# Exercise 2
A store has requested a manufacturer to produce pants and sports jackets. For materials, the manufacturer has 750𝑚2 of cotton textile and 1000𝑚2 of polyester. Every pair of pants (1 unit) needs 1𝑚2 of cotton and 2𝑚2 of polyester. Every jacket needs 1.5𝑚2 of cotton and 1𝑚2 of polyester. The price of the pants is fixed at 50€ and the jacket, 40€. What is the number of pants and jackets that the manufacturer must give to the stores so that these items obtain a maximum sale?

Formulating as a linear programming problem: let $x_1$, $x_2$ be numbers of jeans and jackets. \
Then: \
max z(x) =    50$x_1$ + 40$x_2$             (0) - maximizing profit from items \
                      $x_1$ + 1.5$x_2$ <= 750     (1) - available $m^2$ of cotton textile \
                      2$x_1$ + $x_2$ <= 1000     (2) - available $m^2$ of polyester \
                      $x_1$, $x_2$ >= 0                (3) - numbers can't be negative and must be integers \
                      $x_1, x_2 \in \mathbb{N}$


In [69]:
model2 = Model(GLPK.Optimizer)

A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: GLPK

In [70]:
@variable(model2, x1 >= 0, Int)
@variable(model2, x2 >= 0, Int)

x2

In [72]:
@objective(model2, Max, 50x1 + 40x2)

50 x1 + 40 x2

In [73]:
@constraint(model2, cst1, x1 + 1.5x2 <= 750)
@constraint(model2, cst2, 2x1 + x2 <= 1000)

cst2 : 2 x1 + x2 ≤ 1000

In [74]:
print(model2)

In [75]:
optimize!(model2)

In [76]:
@show termination_status(model2)
@show primal_status(model2)

termination_status(model2) = MathOptInterface.OPTIMAL
primal_status(model2) = MathOptInterface.FEASIBLE_POINT


FEASIBLE_POINT::ResultStatusCode = 1

In [78]:
@show objective_value(model2)
@show value(x1)
@show value(x2)

objective_value(model2) = 28750.0
value(x1) = 375.0
value(x2) = 250.0


250.0

In [80]:
solution_summary(model2)

* Solver : GLPK

* Status
  Result count       : 1
  Termination status : OPTIMAL
  Message from the solver:
  "Solution is optimal"

* Candidate solution (result #1)
  Primal status      : FEASIBLE_POINT
  Dual status        : NO_SOLUTION
  Objective value    : 2.87500e+04
  Objective bound    : 2.87500e+04
  Relative gap       : 0.00000e+00

* Work counters
  Solve time (sec)   : 1.71494e-03


## Answer
The manufacturer must give to the stores 375 jeans and 250 jackets. \
The overall income will be 28750 Euros.

# Exercise 3
MetaCorp is considering four investments. Investment 1 will yield a net present value (NPV) of 16000€; investment 2, an NPV of 22000€; investment 3, an NPV of 12000€; and investment 4, an NPV of 8000€. Each investment requires a certain cash outflow at the present time: investment 1, 5000€; investment 2, 7000€; investment 3, 4000€; and investment 4, 3000€. At present, 14000€ is available for investment. Formulate a linear programming problem whose solution will tell MetaCorp how to maximize the NPV obtained from investiments 1 − 4.

Formulating as a linear programming problem: let $x_1, x_2, x_3, x_4$ be the signals if investment was made or not. \
Then: \
max z(x) =    $16000x_1 + 22000x_2 + 12000x_3 + 8000x_4$                     (0) - maximizing profit from investments \
                      $5000x_1 + 7000x_2 + 4000x_3 + 3000x_4 <= 14000$       (1) - available funds for invertments now \
                      $x_1, x_2, x_3, x_4 \in \{0; 1\}$                                                         (2) - investment can be made or not and only once

Let's make an implicit model this time as there are more variables. \
c = (16000, 22000, 12000, 8000) \
d = (14) \
T = (5000, 7000, 4000, 3000)

In [98]:
c = [16000, 22000, 12000, 8000]
d = [14000]
T = [5000 7000 4000 3000]
n, m = size(T)

(1, 4)

In [99]:
model3 = Model(GLPK.Optimizer)
@variable(model3, x[1:m], Bin)
@objective(model3, Max, sum(c[j]*x[j] for j=1:m))
@constraint(model3, cst[i=1:n], sum(T[i,j]*x[j] for j=1:m) <= d[i])

1-element Vector{ConstraintRef{Model, MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.LessThan{Float64}}, ScalarShape}}:
 cst[1] : 5000 x[1] + 7000 x[2] + 4000 x[3] + 3000 x[4] ≤ 14000

In [100]:
print(model3)

In [101]:
optimize!(model3)

In [102]:
@show termination_status(model3)
@show primal_status(model3)

termination_status(model3) = MathOptInterface.OPTIMAL
primal_status(model3) = MathOptInterface.FEASIBLE_POINT


FEASIBLE_POINT::ResultStatusCode = 1

In [103]:
@show objective_value(model3)
@show value.(x)

objective_value(model3) = 42000.0
value.(x) = [0.0, 1.0, 1.0, 1.0]


4-element Vector{Float64}:
 0.0
 1.0
 1.0
 1.0

In [106]:
solution_summary(model3)

* Solver : GLPK

* Status
  Result count       : 1
  Termination status : OPTIMAL
  Message from the solver:
  "Solution is optimal"

* Candidate solution (result #1)
  Primal status      : FEASIBLE_POINT
  Dual status        : NO_SOLUTION
  Objective value    : 4.20000e+04
  Objective bound    : 4.20000e+04
  Relative gap       : 0.00000e+00

* Work counters
  Solve time (sec)   : 1.31130e-04


## Answer 
MetaCorp should make Investments # 2, 3, 4 \
They will recieve 42000 Euros.